gateway/gateway_test.go

```diff
@@ -39,3 +42,198 @@ func TestValidate(t *testing.T) {
 	_, err = conf.Validate("")
 	test.That(t, err, test.ShouldBeError, resource.NewConfigValidationError("", errInvalidSpiBus))
 }
+
+func TestDoCommand(t *testing.T) {
+	g := &Gateway{
+		devices: make(map[string]*node.Node),
+	}
+
+	s := (sensor.Sensor)(g)
+
+	n := node.Node{
+		AppKey:      testAppKey,
+		DevEui:      testDevEUI,
+		DecoderPath: testDecoderPath,
+		NodeName:    testNodeName,
+		JoinType:    "OTAA",
+	}
+
+	// Test Validate response should be 1
+	validateCmd := make(map[string]interface{})
+	validateCmd["validate"] = 1
+	resp, err := s.DoCommand(context.Background(), validateCmd)
+	test.That(t, err, test.ShouldBeNil)
+	retVal, ok := resp["validate"]
+	test.That(t, ok, test.ShouldBeTrue)
+	test.That(t, retVal.(int), test.ShouldEqual, 1)
+
+	// need to simulate what happens when the DoCommand message is serialized/deserialized into proto
+	doOverWire := func(gateway sensor.Sensor, cmd map[string]interface{}) {
+		command, err := protoutils.StructToStructPb(cmd)
+		test.That(t, err, test.ShouldBeNil)
+		_, err = gateway.DoCommand(context.Background(), command.AsMap())
+		test.That(t, err, test.ShouldBeNil)
+	}
+
+	// Test Register device command
+	registerCmd := make(map[string]interface{})
+	registerCmd["register_device"] = n
+	doOverWire(s, registerCmd)
+	test.That(t, len(g.devices), test.ShouldEqual, 1)
+	dev, ok := g.devices[testNodeName]
+	test.That(t, ok, test.ShouldBeTrue)
+	test.That(t, dev.AppKey, test.ShouldResemble, testAppKey)
+	test.That(t, dev.DevEui, test.ShouldResemble, testDevEUI)
+	test.That(t, dev.DecoderPath, test.ShouldResemble, testDecoderPath)
+
+	// Test that if the device reconfigures, fields get updated
+	n.DecoderPath = "/newpath"
+	registerCmd["register_device"] = n
+	doOverWire(s, registerCmd)
+	test.That(t, len(g.devices), test.ShouldEqual, 1)
+	dev, ok = g.devices[testNodeName]
+	test.That(t, ok, test.ShouldBeTrue)
+	test.That(t, dev.AppKey, test.ShouldResemble, testAppKey)
+	test.That(t, dev.DevEui, test.ShouldResemble, testDevEUI)
+	test.That(t, dev.DecoderPath, test.ShouldResemble, "/newpath")
+
+	// Test remove device command
+	removeCmd := make(map[string]interface{})
+	removeCmd["remove_device"] = n.NodeName
+	doOverWire(s, removeCmd)
+	test.That(t, len(g.devices), test.ShouldEqual, 0)
```

Comment:
Can we test some error cases of your DoCommands, if there are any in this testing function?